certora/specs/Reverts.spec

```diff
@@ -0,0 +1,256 @@
+    function MORPHO.market(PreLiquidation.Id) external
+      returns (uint128, uint128, uint128,uint128, uint128, uint128) envfree;
+    function MORPHO.position(PreLiquidation.Id, address) external
+      returns (uint256, uint128, uint128) envfree;
+    function MORPHO.idToMarketParams(PreLiquidation.Id) external
+      returns (address, address, address, address, uint256) envfree;
+    function _.price() external => mockPrice() expect uint256;
+}
+
+persistent ghost uint256 lastPrice;
+persistent ghost bool priceChanged;
+persistent  ghost bool preLiquidateReverted;
+
+function mockPrice() returns uint256 {
+    uint256 updatedPrice;
+    if (updatedPrice != lastPrice) {
+        priceChanged = true;
+        lastPrice = updatedPrice;
+    }
+    return updatedPrice;
+}
+
+function summaryMulDivUp(uint256 x,uint256 y, uint256 d) returns uint256 {
+    // Safe require because the reference implementation would revert.
+    return require_uint256((x * y + (d-1)) / d);
+}
+
+function summaryMulDivDown(uint256 x,uint256 y, uint256 d) returns uint256 {
+    // Safe require because the reference implementation would revert.
+    return require_uint256((x * y)/d);
+}
+
+definition WAD() returns uint256 = 10^18;
+
+definition ORACLE_SCALE() returns uint256  = 10^36;
+
+definition exactlyOneZero(uint256 assets, uint256 shares) returns bool =
+  (assets == 0 && shares != 0) || (assets != 0 && shares == 0);
+
+definition tAU(uint256 shares, uint256 totalAssets, uint256 totalShares) returns uint256 =
+    summaryMulDivUp(shares, require_uint256(totalAssets + (10^6)), require_uint256(totalShares + (10^6)));
+
+definition tSU(uint256 assets, uint256 totalAssets, uint256 totalShares) returns uint256 =
+    summaryMulDivUp(assets, require_uint256(totalShares + (10^6)), require_uint256(totalAssets + (10^6)));
+
+definition wDU(uint256 x,uint256 y) returns uint256 = summaryMulDivUp(x, WAD(), y);
+
+definition wDD(uint256 x,uint256 y) returns uint256 = summaryMulDivDown(x, WAD(), y);
+
+definition wMD(uint256 x,uint256 y) returns uint256 = summaryMulDivDown(x, y, WAD());
+
+definition computeFactor(mathint ltv, mathint lLtv, mathint preLLTV, mathint factor1, mathint factor2)
+    returns mathint =
+    wMD(wDD(require_uint256(ltv - preLLTV),
+            require_uint256(lLtv - preLLTV)),
+        require_uint256(factor2 - factor1)) + factor1;
+
+// Checks that onMorphoRepay is only triggered by Morpho
+rule onMorphoRepaySenderValidation(env e, uint256 repaidAssets, bytes data) {
+    onMorphoRepay@withrevert(e, repaidAssets, data);
+    assert e.msg.sender != currentContract.MORPHO => lastReverted;
+}
+
+function lastUpdateIsNotNil(PreLiquidation.Id id) returns bool {
+    mathint lastUpdate;
+    (_,_,_,_,lastUpdate,_) = MORPHO.market(id);
+    return lastUpdate != 0;
+}
+
+//Ensure constructor requirements.
+
+invariant marketExists()
+    lastUpdateIsNotNil(currentContract.ID);
+
+invariant preLltvLTlltv()
```

Comment:
> I like the idea, how could we make it coherent with respect to markets created?

What do you mean by making it coherent ?

> Also, these invariant are required in later rules to avoid division by zero in summaries.

That's fine, we would just import that ConsistentSetup file in the header of the Reverts file

Reply:
> What do you mean by making it coherent ?
What I had in mind is that, these invariant are actually implying that any instantiation of `PreLiquidation` that falsifies them will trigger a revert.

Reply:
Also, we would have to ensure that `parametric_contracts` setup is the same one in both configurations, right?

Reply:
> What I had in mind is that, these invariant are actually implying that any instantiation of PreLiquidation that falsifies them will trigger a revert.

Yes that's true, but why do you think we have to do something else (when you say "how could we make it coherent with respect to markets created?")

> Also, we would have to ensure that `parametric_contracts` setup is the same one in both configurations, right?

No I don't think so, if the invariants in one file and the invariants on the other file do not depend on the same states

Reply:
> Do you mean they will just be assumed in the other file?

Oh ok so now I get why you said that we would need the same parametric_contracts setting. Yes it's possible to make so they are assumed in the other file, but I don't remember if you need to do something particular for that as the behavior was changed I believe

Reply:
Implemented in https://github.com/morpho-org/pre-liquidation/pull/37/commits/4b1dc0a6ac503942c35ac02ba1cccc52990fea4d.